src/i_3dsound.c

```diff
@@ -0,0 +1,306 @@
+#define FIXED_TO_ALFLOAT(x) ((ALfloat)((double)(x) / FRACUNIT))
+
+typedef struct oal_listener_params_s
+{
+    ALfloat position[3];
+    ALfloat velocity[3];
+    ALfloat orientation[6];
+} oal_listener_params_t;
+
+typedef struct oal_source_params_s
+{
+    ALfloat position[3];
+    ALfloat velocity[3];
+    boolean point_source;
+    fixed_t z;
+} oal_source_params_t;
+
+static void CalcPitchAngle(const player_t *player, angle_t *pitch)
+{
+    fixed_t lookdir, slope;
+
+    // Same as R_SetupFrame() in r_main.c.
+    lookdir = player->lookdir / MLOOKUNIT + player->recoilpitch;
+
+    if (lookdir == 0)
+    {
+        *pitch = 0;
+        return;
+    }
+
+    if (lookdir > LOOKDIRMAX)
+        lookdir = LOOKDIRMAX;
+    else if (lookdir < -LOOKDIRMAX)
+        lookdir = -LOOKDIRMAX;
+
+    // Flip sign due to right-hand rule.
+    slope = (-lookdir << FRACBITS) / VIEWDIST;
+    if (slope < 0)
+    {
+        *pitch = (ANGLE_MAX - tantoangle[-slope >> DBITS]) >> ANGLETOFINESHIFT;
+    }
+    else
+    {
+        *pitch = tantoangle[slope >> DBITS] >> ANGLETOFINESHIFT;
+    }
+}
+
+static void CalcListenerParams(const mobj_t *listener, oal_listener_params_t *lis)
+{
+    const player_t *player = listener->player;
+    const angle_t yaw = listener->angle >> ANGLETOFINESHIFT;
+    angle_t pitch;
+
+    lis->position[0] = FIXED_TO_ALFLOAT(listener->x);
+    lis->position[1] = FIXED_TO_ALFLOAT(player->viewz);
+    lis->position[2] = FIXED_TO_ALFLOAT(-listener->y);
+
+    if (oal_use_doppler)
+    {
+        lis->velocity[0] = FIXED_TO_ALFLOAT(listener->momx) * TICRATE;
+        lis->velocity[1] = FIXED_TO_ALFLOAT(listener->momz) * TICRATE;
+        lis->velocity[2] = FIXED_TO_ALFLOAT(-listener->momy) * TICRATE;
+    }
+    else
+    {
+        lis->velocity[0] = 0.0f;
+        lis->velocity[1] = 0.0f;
+        lis->velocity[2] = 0.0f;
+    }
+
+    CalcPitchAngle(player, &pitch);
+    if (pitch == 0)
+    {
+        // "At" vector after yaw rotation.
+        lis->orientation[0] = FIXED_TO_ALFLOAT(finecosine[yaw]);
+        lis->orientation[1] = 0.0f;
+        lis->orientation[2] = FIXED_TO_ALFLOAT(-finesine[yaw]);
+
+        // "Up" vector doesn't change.
+        lis->orientation[3] = 0.0f;
+        lis->orientation[4] = 1.0f;
+        lis->orientation[5] = 0.0f;
+    }
+    else
+    {
+        // "At" vector after yaw and pitch rotations.
+        lis->orientation[0] = FIXED_TO_ALFLOAT(FixedMul(finecosine[yaw], finecosine[pitch]));
+        lis->orientation[1] = FIXED_TO_ALFLOAT(-finesine[pitch]);
+        lis->orientation[2] = FIXED_TO_ALFLOAT(-FixedMul(finesine[yaw], finecosine[pitch]));
+
+        // "Up" vector after yaw and pitch rotations.
+        lis->orientation[3] = FIXED_TO_ALFLOAT(FixedMul(finecosine[yaw], finesine[pitch]));
+        lis->orientation[4] = FIXED_TO_ALFLOAT(finecosine[pitch]);
+        lis->orientation[5] = FIXED_TO_ALFLOAT(-FixedMul(finesine[yaw], finesine[pitch]));
+    }
```

Comment:
For future reference: [Standard rotation matrix](https://en.wikipedia.org/wiki/Rotation_matrix#General_rotations), α = yaw, β = pitch, ignore roll.

"At" vector in Doom space:
![at_1](https://github.com/fabiangreffrath/woof/assets/56656010/ddf593ee-7144-4d6f-a4f5-f0dcd31565de)

"Up" vector in Doom space:
![up_1](https://github.com/fabiangreffrath/woof/assets/56656010/4ada77b0-5420-4991-8e66-1412a81c313f)

Doom to OpenAL space: `{x, y, z}` to `{x, z, -y}`

Reply:
> Doom to OpenAL space: `{x, y, z}` to `{x, z, -y}`

Could you please add this as a comment somewhere?

Reply:
Added here: https://github.com/fabiangreffrath/woof/pull/1118/commits/a2928faf4fb91ac19897b1705b0f4e5a4e71a3d0